lib/Translation_unit.ml

```diff
@@ -297,86 +299,108 @@ let format xunit ?output_file ~input_name ~source ~parsed conf opts =
+             t_new)
+      then
+        let old_ast = dump_ast ~suffix:".old" (xunit.normalize conf t) in
+        let new_ast = dump_ast ~suffix:".new" (xunit.normalize conf t_new) in
+        let args ~suffix =
+          [ ("output file", dump_formatted ~suffix fmted)
+          ; ("old ast", old_ast)
+          ; ("new ast", new_ast) ]
+          |> List.filter_map ~f:(fun (s, f_opt) ->
+                 Option.map f_opt ~f:(fun f -> (s, String.sexp_of_t f)))
+        in
+        if xunit.equal ~ignore_doc_comments:true conf t t_new then
+          let docstrings = xunit.moved_docstrings conf t t_new in
+          let args = args ~suffix:".unequal-docs" in
+          internal_error (`Doc_comment docstrings) args
+        else
+          let args = args ~suffix:".unequal-ast" in
+          internal_error `Ast_changed args ) ;
+      (* Comments not preserved ? *)
+      if conf.comment_check then (
+        ( match Cmts.remaining_comments cmts_t with
+        | [] -> ()
+        | l -> internal_error (`Comment_dropped l) [] ) ;
+        let is_docstring Cmt.{txt; _} =
+          conf.parse_docstrings && Char.equal txt.[0] '*'
+        in
+        let old_docstrings, old_comments =
+          List.partition_tf t.comments ~f:is_docstring
+        in
+        let t_newdocstrings, t_newcomments =
+          List.partition_tf t_new.comments ~f:is_docstring
+        in
+        let f = ellipsis_cmt in
+        let f x = Either.First.map ~f x |> Either.Second.map ~f in
+        let diff_cmts =
+          Sequence.append
+            (Cmts.diff conf old_comments t_newcomments)
+            (Fmt_odoc.diff conf old_docstrings t_newdocstrings)
+          |> Sequence.map ~f
+        in
+        if not (Sequence.is_empty diff_cmts) then
+          let old_ast = dump_ast ~suffix:".old" t.ast in
+          let new_ast = dump_ast ~suffix:".new" t_new.ast in
+          let args =
+            [ ( "diff"
+              , Some
+                  (Sequence.sexp_of_t
+                     (Either.sexp_of_t String.sexp_of_t String.sexp_of_t)
+                     diff_cmts) )
+            ; ("old ast", Option.map old_ast ~f:String.sexp_of_t)
+            ; ("new ast", Option.map new_ast ~f:String.sexp_of_t) ]
+            |> List.filter_map ~f:(fun (s, f_opt) ->
+                   Option.map f_opt ~f:(fun f -> (s, f)))
+          in
+          internal_error `Comment args ) ;
+      (* Too many iteration ? *)
+      if i >= conf.max_iters then (
+        Caml.flush_all () ;
+        Error (Unstable {iteration= i; prev= source; next= fmted}) )
+      else
+        (* All good, continue *)
+        print_check ~i:(i + 1) ~conf t_new ~source:fmted
   in
   try print_check ~i:1 ~conf parsed ~source with
   | Sys_error msg -> Error (User_error msg)
   | exn -> Error (Ocamlformat_bug {exn})
 
+let parse_result xunit conf (opts : Conf.opts) ~source ~input_name =
+  match parse xunit.parse conf ~source with
+  | exception exn ->
```

Comment:
Looking at `format` above, `parse` could raise several kinds of exceptions:
```
      | exception Sys_error msg -> Error (User_error msg)
      | exception Warning50 l -> internal_error (`Warning50 l) (exn_args ())
      | exception exn -> internal_error (`Cannot_parse exn) (exn_args ())
```
The corresponding errors are no longer reported correctly. I'm not sure it makes sense to apply recovering when the first two (`Sys_error`, `Warning50`) happen.